compiler/rustc_parse/src/parser/generics.rs

```diff
@@ -286,7 +304,74 @@ impl<'a> Parser<'a> {
                     },
                 ));
             } else if self.check_type() {
-                where_clause.predicates.push(self.parse_ty_where_predicate()?);
+                // Save the state of the parser before trying to parse a predicate, in case we fail
```

Comment:
Can we also move all in this block to its own function? Maybe `parse_ty_where_predicate_with_invalid_tuple_struct_recovery` or something :)